CASSANDRA-17044: Refactor schema management

jacek-lewandowski opened the pull request.

jacek-lewandowski committed:
CASSANDRA-17044: Refactor MigrationCoordinator

Make most of the methods private, refactor tests to properly use mocking external services
Moved some static methods from MigrationManager to MigrationCoordinator and made them instance methods, as well as to the test code
It turned out that many static methods in MigrationManager are actually used only in the test code. I created SchemaTestUtil class in test sources and moved those methods there so that MigrationManager is not polluted.

jacek-lewandowski committed:
CASSANDRA-17044: Extracted TableMetadataRefCache

Table metadata reference cache has been extracted out from SchemaManager as a distinct entity.

jacek-lewandowski committed:
CASSANDRA-17044: Extract schema event notifications to SchemaChangeNotifier

- Converted SchemaChangeListener to an interface with default empty methods - refactored overrides and added @Override annotations.
- Extracted functionality of registering and unregistering listeners as well as issuing individual notifications for each updated element in the schema to SchemaChangeNotifier. The only public methods are for whole keyspace creation, dropping and alteration. This simplifies SchemaManager class by removing a lot of code.

jacek-lewandowski committed:
CASSANDRA-17044: Extracted MigrationSerializer to a separate class

jacek-lewandowski committed:
CASSANDRA-17044: Extracted SharedSchema object and SchemaTransformationResult

Static subclass TransformationResult was removed from SchemaManager and added to SchemaTransformation class.

Also removed fields like success or exception as it made no sense for them to be in that class.

jacek-lewandowski committed:
CASSANDRA-17044: Fix confused ring and schema delays in StorageService

Renamed values denoting timeouts so that it is clear which time unit they expect.

jacek-lewandowski committed:
CASSANDRA-17044: Some small fixes around managing Keyspace instances

jacek-lewandowski committed:
CASSANDRA-17044: Fix some typos in SchemaDiagnostics

jacek-lewandowski committed:
CASSANDRA-17044: Some minor additions and fixes

Added a couple of useful methods and made some small self-explanatory fixes.

jacek-lewandowski committed:
CASSANDRA-17044: Fix changes application order in SchemaKeyspace

Sometimes it happens we replace the keyspace. In this case the keyspace with the same name is in dropped and created sets. If we first try to create it, we would just replace the existing one and the remove it, which is not what we want.

jacek-lewandowski committed:
CASSANDRA-17044: Fix post joining notification

Once joined the ring, StorageService artificially executes onChange (endpoint event subscriber method) with all the updated values. However, we want that all registered endpoint event subscribers get updated.

jacek-lewandowski committed:
CASSANDRA-17044: Add pre schema update notifications

jacek-lewandowski committed:
CASSANDRA-17044: Refactor schema updates

The code which is responsible for dealing with SchemaKeyspace (storing changes in system tables) has been extracted out to DefaultSchemaUpdateHandler.
The other changes are the consequences of that change.

Delegate some functionality related to MigrationCoordinator through DefaultSchemaUpdateHandler
DefaultSchemaUpdateHandler also registers itself as a endpoint event subscriber, so we may avoid explicit calls on MigrationCoordinator when certain events happen.

Squash SchemaManager.loadFromDisk methods

Make some methods private in SchemaManager

Remove direct usages of MigrationCoordinator
After previous changes, it became possible to remove some methods and reduce visibility of others. The main change is that we no longer need to get direct MigrationCoordinator instance outside DefaultSchemaUpdateHandler.

jacek-lewandowski committed:
CASSANDRA-17044: Add SchemaUpdateHandlerFactoryProvider

Factory is used to create the right instance depending on the context - whether we are online or not (server mode / tool or client mode). Provider is for creating the factory so that we can switch the implementation by providing a system property. This can be also useful for testing.

jacek-lewandowski committed:
CASSANDRA-17044: Extracted LocalKeyspaces class

LocalKeyspaces holds metadata of the system local keyspaces, namely system and system_schema.
Metadata of those keyspaces are immutable, they are not synced with other nodes and not stored. As such, we want to separate them from metadata of the keyspaces which are modifiable and synced between nodes, hence called sharedKeyspaces.

There a couple of minor refactorings following that change - some query methods in SchemaManager were renamed and refactored to return immutable Keyspaces object rather than ordinary collections of KeyspaceMetadata or names.

jacek-lewandowski committed:
CASSANDRA-17044: Remove schema mutations from transformation result

jtgrabowski reviewed (approved):
+1 (as in review in our repo)

jacek-lewandowski committed:
post rebase - fixes

jacek-lewandowski committed:
post rebase - let Schema be more testable

especially allow for testing Schema without initializing Keyspace / ColumnFamilyStores

jacek-lewandowski committed:
post rebase - allow for removing tables without dropping data

jacek-lewandowski committed:
post rebase - fix Guardrails enabled check

It threw NPE when DD was initialized as client because of the wrong order of checks

jacek-lewandowski committed:
DO NOT MERGE - CircleCI config

jacek-lewandowski committed:
proposed fix for populating system schema keyspace
